mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/notation/FenNotationCombinators.kt

```diff
@@ -0,0 +1,154 @@
+        val rank = requireNotNull(LettersToRank[it.lowercaseChar()])
+        val color = if (it.isUpperCase()) White else Black
+        Square.Piece(rank, color)
+      }
+
+  /**
+   * A [Parser] which consumes digit in FEN representing empty squares and returns the corresponding
+   * [Square.Empty]
+   */
+  private fun empty(): Parser<String, Square.Empty> =
+      digit().filter { it in 1..Board.Size }.map { Square.Empty(it) }
+
+  /**
+   * A [Parser] which consumes either a piece or a digit in FEN and returns the corresponding
+   * [Square]
+   */
+  private fun square(): Parser<String, Square> = piece().or(empty())
+
+  /** A [Parser] which consumes a FEN row delimiter */
+  private fun delimiter(): Parser<String, Unit> = char(RowSeparatorSymbol).map {}.orElse {}
+
+  /**
+   * A [Parser] which consumes an entire FEN row representation, and returns the corresponding
+   * [List] of [Square]s
+   */
+  private fun lineSquares(): Parser<String, List<Square>> =
+      square()
+          .repeatAtLeast(count = 1) // Parse at least one square.
+          .flatMap { list -> delimiter().map { list } }
+
+  /**
+   * A [Parser] which consumes a whole board representation in FEN and returns the corresponding
+   * [List] of [List]s of [Square]s
+   */
+  private fun boardSquares(): Parser<String, List<List<Square>>> = lineSquares().repeat()
+
+  /**
+   * A [Parser] which consumes a board representation in FEN and returns the corresponding [Board]
+   */
+  fun board(): Parser<String, Board<Piece<Color>>> =
+      boardSquares().map { lines ->
+        buildBoard {
+          var id = PieceIdentifier(0)
+          for ((y, line) in lines.withIndex()) {
+            var progress = 0
+            for (square in line) {
+              if (square is Square.Piece) {
+                val position = Position(x = progress, y = y)
+                val piece = Piece(square.color, square.rank, id++)
+                set(position, piece)
+              }
+              progress += square.width
+            }
+          }
+        }
+      }
+
+  /** A [Parser] which consumes either a w or b indicating the next playing color */
+  val activeColor = char('w').map { White } or char('b').map { Black }
+
+  /**
+   * Parses the first [Char] of a [String], if it's not empty and has the provided value whether it
+   * is lower-case of upper-case.
+   *
+   * @param value the value that is searched.
+   */
+  private fun charLower(value: Char): Parser<String, Char> =
+      char().filter { it.lowercaseChar() == value }
+
+  /**
+   * A [Parser] which consumes half of the castling rights in FEN notation and returns a
+   * corresponding [Pair] of rights, the first component indicating the kingside rights and the
+   * second one representing the queenside rights, regardless of color
+   */
+  private val castlingRightsHalf =
+      charLower('k').flatMap { charLower('q').map { Pair(first = true, second = true) } } or
```

Comment:
Question : could the sequence ever be `qk` ?

Reply:
No. The order of the castling rights is defined as always being king then queen, for white and then for black. Although `qk` could make sense, it should never happen.